### PR TITLE
Feature/native

### PR DESCRIPTION
**Pull request by mxmCherry**
[OpenRTB Dynamic Native Ads API Specification Version 1.1](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-Native-Ads-Specification-1-1_2016.pdf) implementation

Also, refactored/reorganized tests (without changing fixtures)

### CODE REVIEW

**Comment by mxmCherry** [comment]
Things to check:
- [ ] request - no fields are missing?
- [x] response - no fields are missing?
- [ ] request - numeric types follow root README recommendations?
- [x] response - numeric types follow root README recommendations?
- [ ] request - all non-required fields are `json:"...,omitempty"`?
- [ ] response - all non-required fields are `json:"...,omitempty"`?
- [ ] request - all enums implemented?
- [ ] request - enums naming is OK? (keep it simple - copypaste/camelize, not more)

**Comment by mxmCherry** [comment]
@dbemiller would be great if you'd take a look as well - to spot any obvious mistakes or bad code/naming.

Will do careful review over a few evenings (up to 3 days counting from tomorrow).

**Comment by dbemiller** [comment]
Will do. Thanks!

File: native/request/asset.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	// Field:
+	//   title
+	// Scope:
+	//   recommended (each asset object may contain only one of title, img, data or video)
+	// Type:
+	//   object
+	// Description:
+	//   Title object for title assets.
+	//   See TitleObject definition.
+	Title interface{} `json:"title,omitempty"`
```
Comment (mxmCherry) [inline native/request/asset.go]: `*Title`

File: native/request/asset.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	// Field:
+	//   img
+	// Scope:
+	//   recommended (each asset object may contain only one of title, img, data or video)
+	// Type:
+	//   object
+	// Description:
+	//   Image object for image assets.
+	//   See ImageObject definition.
+	Img interface{} `json:"img,omitempty"`
```
Comment (mxmCherry) [inline native/request/asset.go]: Use *request.Img here

File: native/request/asset.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	// Field:
+	//   video
+	// Scope:
+	//   optional (each asset object may contain only one of title, img, data or video)
+	// Type:
+	//   object - Video object for video assets.
+	//   See the Video request object definition.
+	//   Note that in-stream (ie preroll, etc) video ads are not part of Native.
+	//   Native ads may contain a video as the ad creative itself.
+	Video interface{} `json:"video,omitempty"`
```
Comment (mxmCherry) [inline native/request/asset.go]: Use `*request.Video` here

File: native/request/asset.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	// Field:
+	//   data
+	// Scope:
+	//   recommended (each asset object may contain only one of title, img, data or video)
+	// Type:
+	//   object
+	// Description:
+	//   Data object for brand name, description, ratings, prices etc.
+	//   See DataObject definition.
+	Data interface{} `json:"data,omitempty"`
```
Comment (mxmCherry) [inline native/request/asset.go]: Use `*request.Data` here

File: native/request/image.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+	// Field:
+	//   w
+	// Scope:
+	//   optional
+	// Type:
+	//   integer
+	// Description:
+	//   Width of the image in pixels.
+	W int64 `json:"w,omitempty"`
```
Comment (mxmCherry) [inline native/request/image.go]: `uint64`

File: native/request/image.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	// Scope:
+	//   recommended
+	// Type:
+	//   integer
+	// Description:
+	//   The minimum requested width of the image in pixels.
+	//   This option should be used for any rescaling of images by the client.
+	//   Either w or wmin should be transmitted.
+	//   If only w is included, it should be considered an exact requirement.
+	WMin int64 `json:"wmin,omitempty"`
```
Comment (mxmCherry) [inline native/request/image.go]: `uint64`

File: native/request/image.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+	// Field:
+	//   h
+	// Scope:
+	//   optional
+	// Type:
+	//   integer
+	// Description:
+	//   Height of the image in pixels.
+	H int64 `json:"h,omitempty"`
```
Comment (mxmCherry) [inline native/request/image.go]: `uint64`

File: native/request/image.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	// Scope:
+	//   recommended
+	// Type:
+	//   integer
+	// Description:
+	// The minimum requested height of the image in pixels.
+	// This option should be used for any rescaling of images by the client.
+	// Either h or hmin should be transmitted.
+	// If only h is included, it should be considered an exact requirement.
+	HMin int64 `json:"hmin,omitempty"`
```
Comment (mxmCherry) [inline native/request/image.go]: `uint64`

File: native/request/video.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	// Field:
+	//   protocols
+	// Scope:
+	//   required
+	// Type:
+	//   array of integers
+	// Description:
+	//   An array of video protocols the publisher can accept in the bid response.
+	// See OpenRTB Table ‘Video Bid Response Protocols’ for a list of possible values.
+	Protocols []int8 `json:"protocols"` // TODO!!!
```
Comment (mxmCherry) [inline native/request/video.go]: Copy-paste and use `openrtb.Protocol` here

**Comment by mxmCherry** [comment]
Checked everything scheduled: https://github.com/mxmCherry/openrtb/pull/19#issuecomment-344029337

Quite satisfied with everything, it seems.

Comments are fine, probs, not perfect (like one sentence per line), but that's fine, can be revised later (cosmetics).

Native request enums naming is acceptable. Also, can be renamed later (won't be the best move, but that can be solved with `sed 's/.../.../g'`, so fine. Now focused on proper autocomplete (IDE/editors) and tried not to keep them very long.

Will merge this tomorrow or the day after tomorrow.

**Comment by mxmCherry** [comment]
Nothing really changed in existing code (only tests reorganized + some comments for RawJSON), so it'll become a v9.2.0 release - everything is back-compatible, native bits are not imported/used on openrtb root package (at least yet).

**Comment by dbemiller** [comment]
Thanks again!

Sorry... I got backed up by surprise prod stuff--which I'm still dealing with.

Just wanted to make sure you knew I didn't mean to ignore this... but I haven't had time to look it over. Of course if you're good with it, then that's good enough for me ^^. You've done a great job being careful and consistent so far.

**Comment by mxmCherry** [comment]
Don't bother, you'll test it by using anyway :)

**Comment by mxmCherry** [comment]
Merged, released as [v9.2.0](https://github.com/mxmCherry/openrtb/releases/tag/v9.2.0)

**Pull request merged**